core/macros/src/class.rs

```diff
@@ -0,0 +1,703 @@
+                    boa_engine::js_string!( #name ),
+                    g,
+                    s,
+                    boa_engine::property::Attribute::CONFIGURABLE
+                        | boa_engine::property::Attribute::NON_ENUMERABLE,
+                );
+            }
+        }
+    }
+}
+
+/// A visitor for the `impl X { ... }` block. This will record all top-level functions
+/// and create endpoints for the JavaScript class.
+#[derive(Debug)]
+struct ClassVisitor {
+    // The type name for this class.
+    type_: Type,
+
+    // Whether we detected a constructor while visiting.
+    constructor: Option<Function>,
+
+    // All static functions recorded.
+    statics: Vec<Function>,
+
+    // All methods recorded.
+    methods: Vec<Function>,
+
+    // All accessors (getters and/or setters) with their names.
+    accessors: BTreeMap<String, Accessor>,
+
+    // All errors we found along the way.
+    errors: Option<syn::Error>,
+}
+
+impl ClassVisitor {
+    fn new(type_: Type) -> Self {
+        Self {
+            type_,
+            constructor: None,
+            statics: Vec::new(),
+            methods: Vec::new(),
+            accessors: BTreeMap::default(),
+            errors: None,
+        }
+    }
+
+    fn name_of(fn_: &mut ImplItemFn) -> SpannedResult<String> {
+        take_name_value_attr(&mut fn_.attrs, "name").map_or_else(
+            || Ok(fn_.sig.ident.to_string()),
+            |nv| match &nv {
+                Lit::Str(s) => Ok(s.value()),
+                _ => error(&nv, "Invalid attribute value literal"),
+            },
+        )
+    }
+
+    fn method(&mut self, _span: &impl Spanned, fn_: &mut ImplItemFn) {
+        let name = match Self::name_of(fn_) {
+            Ok(name) => name,
+            Err((span, msg)) => {
+                self.error(span, msg);
+                return;
+            }
+        };
+
+        match Function::method(name, fn_, &self.type_) {
+            Ok(f) => {
+                if f.is_static {
+                    self.methods.push(f);
+                } else {
+                    self.statics.push(f);
+                }
+            }
+            Err((span, err)) => self.error(span, err),
+        }
+    }
+
+    fn getter(&mut self, fn_: &mut ImplItemFn) {
+        let name = match Self::name_of(fn_) {
+            Ok(name) => name,
+            Err((span, msg)) => {
+                self.error(span, msg);
+                return;
+            }
+        };
+        if let Err((span, msg)) =
+            self.accessors
+                .entry(name.clone())
+                .or_default()
+                .set_getter(name, fn_, &self.type_)
+        {
+            self.error(span, msg);
+        }
+    }
+
+    fn setter(&mut self, fn_: &mut ImplItemFn) {
+        let name = match Self::name_of(fn_) {
+            Ok(name) => name,
+            Err((span, msg)) => {
+                self.error(span, msg);
+                return;
+            }
+        };
```

Comment:
Perhaps this should be moved into `name_of` to avoid the repetition

Reply:
Issue is the `return` there. I can refactor to have `Result<>` all the way.